typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1507 @@
+  end
+
+  type t
+  (** Mapping from identifiers to a list of possible nodes, each represented by
+      a path into the forest, instead of directly ponting to the node. *)
+
+  val extend : t -> Extension.t -> t
+  (** Extend a mapping with an extension *)
+
+  val empty : t
+  (** The empty mapping  *)
+
+  val find_opt : Ident.t -> t -> UF.Path.t list option
+  (** Find the list of paths corresponding to an identifier  *)
+end = struct
+  module Extension = struct
+    type t = UF.Path.t list Ident.Map.t
+
+    let disjunct ienv0 ienv1 =
+      Ident.Map.merge
+        (fun _id locs0 locs1 ->
+          match (locs0, locs1) with
+          | None, None -> None
+          | Some locs0, Some locs1 -> Some (locs0 @ locs1)
+          (* cannot bind variable only in one of the OR-patterns *)
+          | _, _ -> assert false)
+        ienv0 ienv1
+
+    let empty = Ident.Map.empty
+
+    let conjunct ienv0 ienv1 =
+      Ident.Map.union
+        (fun _id _ _ ->
+          (* cannot bind variable twice in a single pattern *)
+          assert false)
+        ienv0 ienv1
+
+    let conjuncts = List.fold_left conjunct empty
+    let singleton id locs = Ident.Map.singleton id locs
+  end
+
+  type t = UF.Path.t list Ident.Map.t
+
+  let empty = Ident.Map.empty
+
+  let extend t ex =
+    Ident.Map.union
+      (* the extension shadows the original *)
+        (fun _id _paths0 paths1 -> Some paths1)
+      t ex
+
+  let find_opt = Ident.Map.find_opt
+end
+
+(* The fun algebraic stuff ends. Here comes the concrete mess *)
+
+(* Forcing due to boundary is more about OCaml than the algebra, hence defined
+   here (as opposed to earlier) *)
+
+(** See [report_error] for explanation *)
+type boundary_reason =
+  | Value_from_mod_class (* currently will never trigger *)
+  | Free_var_of_mod_class (* currently will never trigger *)
+  | Out_of_mod_class
+
+type boundary = {
+  cannot_force : Maybe_unique.cannot_force;
+  reason : boundary_reason;  (** which kind of boundary is being crossed? *)
+}
+
+exception Boundary of boundary
+
+let force_shared_boundary t ~reason =
+  match Maybe_unique.force_shared t with
+  | Ok () -> ()
+  | Error cannot_force -> raise (Boundary { cannot_force; reason })
+
+type single_value_to_match = {
+  paths : UF.Path.t list;  (** the value's all possible paths. *)
+  loc : Location.t;  (** the location where it's defined *)
+}
+(** Represent a single value that will be pattern-matched upon *)
+
+type value_to_match =
+  | Match_tuple of (single_value_to_match * unique_use option) list
+      (** The value being matched is a tuple; we treat it specially so matching
+  tuples against tuples merely create alias instead of uses; the [unique_use] is
+  used if the tuple is bound to a variable, in which case all values in the tuple
+    is considered used *)
+  | Match_single of single_value_to_match
+      (** The value being matched is not a tuple *)
+
+let mark_implicit_borrow_memory_address_paths access paths loc =
```

Comment:
I think you only need versions of `mark_implicit_borrow_memory_address` that work on the `single_value` and `single_value_use` types.